File: bubus-ts/examples/immediate_event_processing.ts
#!/usr/bin/env -S node --import tsx
// Run: node --import tsx examples/immediate_event_processing.ts

import { z } from 'zod'

import { BaseEvent, EventBus } from '../src/index.js'

// Parent handler runs two scenarios:
// 1) await child.done()               -> immediate queue-jump processing
// 2) await child.waitForCompletion()  -> normal queue processing
const ParentEvent = BaseEvent.extend('ImmediateProcessingParentEvent', {
  mode: z.enum(['immediate', 'queued']),
})

const ChildEvent = BaseEvent.extend('ImmediateProcessingChildEvent', {
  scenario: z.enum(['immediate', 'queued']),
})

const SiblingEvent = BaseEvent.extend('ImmediateProcessingSiblingEvent', {
  scenario: z.enum(['immediate', 'queued']),
})

const delay = (ms: number): Promise<void> =>
  new Promise((resolve) => {
    setTimeout(resolve, ms)
  })

type Scenario = 'immediate' | 'queued'

async function main(): Promise<void> {
  // Two buses: bus_a is the source, bus_b is the forward target.
  const bus_a = new EventBus('QueueJumpDemoA', {
    event_concurrency: 'bus-serial',
    event_handler_concurrency: 'serial',
  })
  const bus_b = new EventBus('QueueJumpDemoB', {
    event_concurrency: 'bus-serial',
    event_handler_concurrency: 'serial',
  })

  // Simple step counter so ordering is easy to read in stdout.
  let step = 0
  const log = (message: string): void => {
    step += 1
    console.log(`${String(step).padStart(2, '0')}. ${message}`)
  }

  // Forwarding setup: both sibling/child events emitted on bus_a are forwarded to bus_b.
  bus_a.on(ChildEvent, (event) => {
    log(`[forward] ${event.event_type}(${event.scenario}) bus_a -> bus_b`)
    bus_b.dispatch(event)
  })
  bus_a.on(SiblingEvent, (event) => {
    log(`[forward] ${event.event_type}(${event.scenario}) bus_a -> bus_b`)
    bus_b.dispatch(event)
  })

  // Local handlers on bus_a.
  bus_a.on(ChildEvent, async (event) => {
    log(`[bus_a] child start (${event.scenario})`)
    await delay(8)
    log(`[bus_a] child end (${event.scenario})`)
  })
  bus_a.on(SiblingEvent, async (event) => {
    log(`[bus_a] sibling start (${event.scenario})`)
    await delay(14)
    log(`[bus_a] sibling end (${event.scenario})`)
  })

  // Forwarded handlers on bus_b.
  bus_b.on(ChildEvent, async (event) => {
    log(`[bus_b] child start (${event.scenario})`)
    await delay(4)
    log(`[bus_b] child end (${event.scenario})`)
  })
  bus_b.on(SiblingEvent, async (event) => {
    log(`[bus_b] sibling start (${event.scenario})`)
    await delay(6)
    log(`[bus_b] sibling end (${event.scenario})`)
  })

  // Parent handler queues sibling first, then child, then compares await behavior.
  bus_a.on(ParentEvent, async (event) => {
    log(`[parent:${event.mode}] start`)

    // Queue a sibling first so normal queue order has sibling ahead of child.
    event.bus?.emit(SiblingEvent({ scenario: event.mode }))
    log(`[parent:${event.mode}] sibling queued`)

    // Queue child second; this is the event we await in two different ways.
    const child = event.bus?.emit(ChildEvent({ scenario: event.mode }))!
    log(`[parent:${event.mode}] child queued`)

    if (event.mode === 'immediate') {
      // Queue-jump: child processes immediately while still inside parent handler.
      log(`[parent:${event.mode}] await child.done()`)
      await child.done()
      log(`[parent:${event.mode}] child.done() resolved`)
    } else {
      // Normal queue wait: child waits its turn behind already-queued sibling work.
      log(`[parent:${event.mode}] await child.waitForCompletion()`)
      await child.waitForCompletion()
      log(`[parent:${event.mode}] child.waitForCompletion() resolved`)
    }

    log(`[parent:${event.mode}] end`)
  })

  const runScenario = async (mode: Scenario): Promise<void> => {
    log(`----- scenario=${mode} -----`)

    // Parent event uses parallel concurrency so waitForCompletion() in handler
    // can wait safely while other queued events continue to run.
    const parent = bus_a.dispatch(
      ParentEvent({
        mode,
        event_concurrency: 'parallel',
      })
    )

    await parent.waitForCompletion()
    await Promise.all([bus_a.waitUntilIdle(), bus_b.waitUntilIdle()])
    log(`----- done scenario=${mode} -----`)
  }

  await runScenario('immediate')
  await runScenario('queued')

  console.log('\nExpected behavior:')
  console.log('- immediate: child runs before sibling (queue-jump) and parent resumes right after child.')
  console.log('- queued: sibling runs first, child waits in normal queue order, parent resumes later.')
  console.log('\n=== bus_a.logTree() ===')
  console.log(bus_a.logTree())
  console.log('\n=== bus_b.logTree() ===')
  console.log(bus_b.logTree())
}

await main()

File: examples/immediate_event_processing.py
#!/usr/bin/env -S uv run python
"""Run: uv run python examples/immediate_event_processing.py"""

import asyncio
from typing import Literal

from bubus import BaseEvent, EventBus, EventConcurrencyMode


class ParentEvent(BaseEvent[None]):
    mode: Literal['immediate', 'queued']


class ChildEvent(BaseEvent[None]):
    scenario: Literal['immediate', 'queued']


class SiblingEvent(BaseEvent[None]):
    scenario: Literal['immediate', 'queued']


async def delay_ms(ms: int) -> None:
    await asyncio.sleep(ms / 1000.0)


async def main() -> None:
    bus_a = EventBus(
        name='QueueJumpDemoA',
        event_concurrency='bus-serial',
        event_handler_concurrency='serial',
    )
    bus_b = EventBus(
        name='QueueJumpDemoB',
        event_concurrency='bus-serial',
        event_handler_concurrency='serial',
    )

    try:
        step = 0

        def log(message: str) -> None:
            nonlocal step
            step += 1
            print(f'{step:02d}. {message}')

        # Forward sibling/child events from bus_a -> bus_b.
        def forward_child(event: ChildEvent) -> None:
            log(f'[forward] {event.event_type}({event.scenario}) bus_a -> bus_b')
            bus_b.emit(event)

        def forward_sibling(event: SiblingEvent) -> None:
            log(f'[forward] {event.event_type}({event.scenario}) bus_a -> bus_b')
            bus_b.emit(event)

        bus_a.on(ChildEvent, forward_child)
        bus_a.on(SiblingEvent, forward_sibling)

        # Local handlers on bus_a.
        async def on_child_a(event: ChildEvent) -> None:
            log(f'[bus_a] child start ({event.scenario})')
            await delay_ms(8)
            log(f'[bus_a] child end ({event.scenario})')

        async def on_sibling_a(event: SiblingEvent) -> None:
            log(f'[bus_a] sibling start ({event.scenario})')
            await delay_ms(14)
            log(f'[bus_a] sibling end ({event.scenario})')

        bus_a.on(ChildEvent, on_child_a)
        bus_a.on(SiblingEvent, on_sibling_a)

        # Forwarded handlers on bus_b.
        async def on_child_b(event: ChildEvent) -> None:
            log(f'[bus_b] child start ({event.scenario})')
            await delay_ms(4)
            log(f'[bus_b] child end ({event.scenario})')

        async def on_sibling_b(event: SiblingEvent) -> None:
            log(f'[bus_b] sibling start ({event.scenario})')
            await delay_ms(6)
            log(f'[bus_b] sibling end ({event.scenario})')

        bus_b.on(ChildEvent, on_child_b)
        bus_b.on(SiblingEvent, on_sibling_b)

        # Parent handler queues sibling first, then child, then compares await behavior.
        async def on_parent(event: ParentEvent) -> None:
            log(f'[parent:{event.mode}] start')

            event.event_bus.emit(SiblingEvent(scenario=event.mode))
            log(f'[parent:{event.mode}] sibling queued')

            child = event.event_bus.emit(ChildEvent(scenario=event.mode))
            log(f'[parent:{event.mode}] child queued')

            if event.mode == 'immediate':
                # Immediate: queue-jump by awaiting child directly inside handler context.
                log(f'[parent:{event.mode}] await child')
                await child
                log(f'[parent:{event.mode}] child await resolved')
            else:
                # Queued: wait on completion signal without queue-jump processing.
                log(f'[parent:{event.mode}] await child.event_completed()')
                await child.event_completed()
                log(f'[parent:{event.mode}] child.event_completed() resolved')

            log(f'[parent:{event.mode}] end')

        bus_a.on(ParentEvent, on_parent)

        async def run_scenario(mode: Literal['immediate', 'queued']) -> None:
            log(f'----- scenario={mode} -----')

            parent = bus_a.emit(
                ParentEvent(
                    mode=mode,
                    event_concurrency=EventConcurrencyMode.PARALLEL,
                )
            )

            await parent
            await asyncio.gather(bus_a.wait_until_idle(), bus_b.wait_until_idle())
            log(f'----- done scenario={mode} -----')

        await run_scenario('immediate')
        await run_scenario('queued')

        print('\nExpected behavior:')
        print('- immediate: child runs before sibling (queue-jump) and parent resumes right after child.')
        print('- queued: sibling runs first, child waits in normal queue order, parent resumes later.')
        print('\n=== bus_a.log_tree() ===')
        print(bus_a.log_tree())
        print('\n=== bus_b.log_tree() ===')
        print(bus_b.log_tree())
    finally:
        await bus_a.stop(clear=True, timeout=0)
        await bus_b.stop(clear=True, timeout=0)


if __name__ == '__main__':
    asyncio.run(main())
